### PR TITLE
[dv] Support data_intg_passthru for sram, rom and flash

### PR DESCRIPTION
**Pull request by weicaiyang**
There are 3 commits in this PR.

Commit 1: [dv] Pass data_intg_passthru to dv_base_mem
Need to know if mem has passthr data_intg, in order to know if we can
skip testing d_user.data_intg error

Commit 2:[dv] Update CIP to handle passthru mem
1. provide a flag to disable checking d_user.data_intg for passthru
memory
2. Remove old similar flag in sram_ctrl and add tl_intg_alert_fields to
fix sram_ctrl_tl_intg_err

Commit 3:Enable data_intg_passthru for rom, sram, flash mem
These mem are meant to have data intg passthru. With that enabled,
data_intg will be generated in ibex and these mem won't re-generate it
again.

Set `disable_d_user_data_intg_check_for_passthru_mem` in
rom_ctrl_env_cfg as the rom isn't initilized with correct ECC

### CODE REVIEW

File: hw/dv/sv/dv_base_reg/dv_base_mem.sv
Comment (rswarbrick) [inline hw/dv/sv/dv_base_reg/dv_base_mem.sv:14]: Nit: **a**long

File: hw/dv/sv/cip_lib/cip_base_env_cfg.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/cip_base_env_cfg.sv:49]: I'm not sure that we should mention Ibex in a CIP base class: the whole point of the comportable IP is that it can be used in other systems. More generally, don't we want to do this unconditionally? I'm not sure why we need this flag?

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/cip_base_scoreboard.sv:462]: It's a pity we have to do the lookup twice here. Can't we just call `ral.default_map.get_mem_by_offset(addr)` and return zero if the result is null?

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv:177]: I'm not a big fan of assigning to the name of the function like this. I think something like this would probably be easier to debug if something goes wrong:
```systemverilog
    bit cmd_intg_ok = !en_cmd_intg || (act_a_user.cmd_intg == exp_a_user.cmd_intg);
    bit data_intg_ok = !en_data_intg || (act_a_user.data_intg == exp_a_user.data_intg);
    bit intg_ok = cmd_intg_ok && data_intg_ok;

    if (!intg_ok) begin
      string str = $sformatf("a_user act (%p) != exp (%p), en_cmd_intg: %0d, en_data_intg: %0d",
                            act_a_user, exp_a_user, en_cmd_intg, en_data_intg);
      if (throw_error) begin
        `uvm_error(`gfn, str)
      end else begin
        `uvm_info(`gfn, str, UVM_MEDIUM)
      end
    end

    return intg_ok;
```
It might also be nice to split out the two fields in the error string, otherwise someone who's debugging is going to have to do the bit twiddling in their head.

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv:196]: Same comment here as for the A channel version above.

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv:171]: I think these arguments should probably be something like `check_cmd_intg`: it's not entirely clear to me what "enable" means in this context.

File: hw/ip/sram_ctrl/dv/env/seq_lib/sram_ctrl_smoke_vseq.sv
Comment (rswarbrick) [inline hw/ip/sram_ctrl/dv/env/seq_lib/sram_ctrl_smoke_vseq.sv:72]: Since you're touching this code anyway, I don't think that "re-init" is a commonly understood shortening of "re-initialising". Maybe expand it?

File: hw/ip/rom_ctrl/data/rom_ctrl.hjson
Comment (rswarbrick) [inline hw/ip/rom_ctrl/data/rom_ctrl.hjson:164]: I would probably prefer to merge https://github.com/lowRISC/opentitan/pull/9539 before making this change. That PR has been waiting for review for a week and fixes the RTL bug properly.

File: hw/ip/rom_ctrl/dv/env/rom_ctrl_env_cfg.sv
Comment (rswarbrick) [inline hw/ip/rom_ctrl/dv/env/rom_ctrl_env_cfg.sv:41]: Again, I'd prefer to merge https://github.com/lowRISC/opentitan/pull/9539 first, in which case we don't need this at all.

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv]: sure, updated. Thanks

**Review by weicaiyang — COMMENTED** [review]
@rswarbrick thanks for the review. I have updated the PR. Please take another look.

File: hw/dv/sv/dv_base_reg/dv_base_mem.sv
Comment (rswarbrick) [reply hw/dv/sv/dv_base_reg/dv_base_mem.sv]: I think this still says "long"

File: hw/dv/sv/cip_lib/cip_base_env_cfg.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_env_cfg.sv]: But do you have any examples where we store memory with integrity and pass it through, but still want to set `data_intg_passthru` to false? I think probably not, which suggests we can drop the flag entirely.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv:463]: That's not what I meant. I meant to do this:
```systemverilog
  virtual function bit is_data_intg_passthru_mem(tl_seq_item item, string ral_name);
    uvm_reg_addr_t addr = cfg.ral_models[ral_name].get_normalized_addr(item.a_addr);
    uvm_mem mem = ral.default_map.get_mem_by_offset(addr);

    if (mem == null) begin
      return 0;
    end else begin
      dv_base_mem dv_mem;
      `downcast(dv_mem, mem)
      return dv_mem.get_data_intg_passthru();
    end
  endfunction
```

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (rswarbrick) [inline hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv:179]: Nit: Missing space before `=`. Also, you are defaulting to `cmd_intg_ok = 0` (unlike what I suggested [here](https://github.com/lowRISC/opentitan/pull/9676#discussion_r768611218)), which means this won't work properly when the check is disabled.

File: hw/dv/sv/dv_base_reg/dv_base_mem.sv
Comment (rswarbrick) [inline hw/dv/sv/dv_base_reg/dv_base_mem.sv:14]: Ah! You fixed this in the wrong commit. Could you grab this hunk for the previous commit instead?

(Git hint: To do this easily, start an interactive rebase, marking the first commit as "edit". Then look up the SHA of this commit and run `git checkout $SHA hw/dv/sv/dv_base_reg/dv_base_mem.sv` to grab the new version of this file. When you continue the rebase, the change will have magically been moved to the first commit).

File: hw/ip/rom_ctrl/data/rom_ctrl.hjson
Comment (rswarbrick) [reply hw/ip/rom_ctrl/data/rom_ctrl.hjson:164]: Thanks!

**Review by matutem — COMMENTED** [review]
I have not much to add after Rupert's review, so I am fine if you commit once he approves.

File: hw/dv/sv/dv_base_reg/dv_base_mem.sv
Comment (weicaiyang) [reply hw/dv/sv/dv_base_reg/dv_base_mem.sv]: I did fix it but in another commit... Sorry about that. It should be right now.

File: hw/dv/sv/cip_lib/cip_base_scoreboard.sv
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_scoreboard.sv]: Thanks, this is better. We can avoid using `is_mem_addr` to look up the mem address again.

File: hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/seq_lib/cip_tl_seq_item.sv]: good catch! fixed them. Thanks

File: hw/dv/sv/dv_base_reg/dv_base_mem.sv
Comment (weicaiyang) [reply hw/dv/sv/dv_base_reg/dv_base_mem.sv:14]: Thanks for the magic tips. It's awesome.

**Comment by rswarbrick** [comment]
I'm still unconvinced that we need the `disable_d_user_data_intg_check_for_passthru_mem` flag at all. Since GitHub's UI isn't super-helpful for finding outstanding comments, it's here: https://github.com/lowRISC/opentitan/pull/9676#discussion_r768604308.

Otherwise, this looks good to me.

File: hw/dv/sv/cip_lib/cip_base_env_cfg.sv
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_env_cfg.sv]: sorry, I missed this comment.

If we always store correct integrity and pass it through, then we don't need this flag. But we also want to store incorrect integrity and see if it's passed through, rather than be re-generated with correct ECC.

For example, there is a negative test for passthru mem #9546. If we flip a bit of the memory content and it's a passthru mem, we should observe an incorrect integrity is returned (this check can be done in seq) and we need to set the flag to disable the error.

Comment (rswarbrick) [reply hw/dv/sv/cip_lib/cip_base_env_cfg.sv]: Ahhh, now I understand! Yes, makes sense.

**Review by rswarbrick — APPROVED** [review]

File: hw/top_earlgrey/data/top_earlgrey.hjson
Comment (tjaychen) [inline hw/top_earlgrey/data/top_earlgrey.hjson:516]: as a separate note, if we add this comment here, we should probably also long term consider how we can pass this directly into the related modules to manipulate the options on the `tlul_adapter_sram`.  Otherwise things could get out of sync.

These tests would catch that though, so probably not a super high priority.

Comment (weicaiyang) [reply hw/top_earlgrey/data/top_earlgrey.hjson:516]: Agree. This is so far used in DV RAL model. Design should pass this to the tlul_adapter_sram from the top.

**Pull request merged**
